File: src/baekJoon/빙산.java
package baekJoon;

// https://www.acmicpc.net/problem/2573

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class 빙산 {

  private static int row, col;
  private static boolean[][] visited;
  private static Iceberg[][] icebergs;
  private static int[] dx = {-1, 1, 0, 0};
  private static int[] dy = {0, 0, -1, 1};

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    StringTokenizer st = new StringTokenizer(br.readLine());

    row = Integer.parseInt(st.nextToken());
    col = Integer.parseInt(st.nextToken());
    icebergs = new Iceberg[row][col];
    visited = new boolean[row][col];

    for(int i = 0; i < row; i++) {
      for(int j = 0; j < col; j++) {
        icebergs[i][j] = new Iceberg();
      }
    }

    for(int i = 0; i < row; i++) {
      st = new StringTokenizer(br.readLine());
      for(int j = 0; j < col; j++) {
        int ice = Integer.parseInt(st.nextToken());
        icebergs[i][j].x = i;
        icebergs[i][j].y = j;
        icebergs[i][j].height = ice;
        if(ice == 0) {
          for(int z = 0; z < 4; z++) {
            int nx = i + dx[z];
            int ny = j + dy[z];
            if(nx < 0 || nx >= row || ny < 0 || ny >= col) continue;
            icebergs[nx][ny].borderOnSeaCnt++;
          }
        }
      }
    }

    int time = -1;
    int chunk = 0;
    while(true) {
      visited = new boolean[row][col];
      chunk = 0;
      for(int i = 0; i < row; i++) {
        for(int j = 0; j < col; j++) {
          if(icebergs[i][j].height != 0 && !visited[i][j]) {
            melt(i, j);
            chunk++;
          }
        }
      }

      time++;
      if(chunk == 0) {
        time = 0;
        break;
      }
      if(chunk >= 2) break;
    }

    bw.write(String.valueOf(time));
    bw.newLine();
    bw.flush();
    bw.close();
    br.close();

  }

  private static void melt(int x, int y) {
    Queue<Iceberg> icebergQueue = new LinkedList<>();
    Queue<Node> removeIce = new LinkedList<>();
    icebergQueue.add(icebergs[x][y]);
    visited[x][y] = true;

    while(!icebergQueue.isEmpty()) {

      Iceberg here = icebergQueue.poll();
      int c_x = here.x;
      int c_y = here.y;
      int c_height = here.height;
      int c_seaCnt = here.borderOnSeaCnt;

      int n_height = Math.max(c_height - c_seaCnt, 0);
      here.height = n_height;
      if(n_height == 0) removeIce.add(new Node(c_x, c_y));

      for(int i = 0; i < 4; i++) {
        int nx = c_x + dx[i];
        int ny = c_y + dy[i];
        if(nx < 0 || nx >= row || ny < 0 || ny >= col) continue;
        if(visited[nx][ny]) continue;
        if(icebergs[nx][ny].height == 0) continue;

        visited[nx][ny] = true;
        icebergQueue.add(icebergs[nx][ny]);
      }
    }
    while (!removeIce.isEmpty()) {
      Node remove = removeIce.poll();
      for(int i = 0; i < 4; i++) {
        int nx = remove.x + dx[i];
        int ny = remove.y + dy[i];
        if(nx < 0 || nx >= row || ny < 0 || ny >= col) continue;
        icebergs[nx][ny].borderOnSeaCnt++;
      }
    }
  }

  private static class Node {
    int x, y;

    public Node(int x, int y) {
      this.x = x;
      this.y = y;
    }
  }

  private static class Iceberg {
    int x, y;
    int height;
    int borderOnSeaCnt;

    public Iceberg() {
      this.height = 0;
      this.borderOnSeaCnt = 0;
    }
  }

}

File: src/kotlin/bfs/빙산.kt
import java.io.BufferedWriter
import java.io.OutputStreamWriter
import java.util.Queue

private val dx = intArrayOf(-1, 1, 0, 0)
private val dy = intArrayOf(0, 0, -1, 1)

fun main() = with(System.`in`.bufferedReader()) {

    val bw = BufferedWriter(OutputStreamWriter(System.out))

    var year = 0

    val (x, y) = readLine().split(" ").map { it.toInt() }
    val ice = Array(x) { IntArray(y) }
    val iceQueue = ArrayDeque<Pair<Int, Int>>()

    for (i in 0 until x) {
        val info = readLine().split(" ").map { it.toInt() }
        for (j in 0 until y) {
            ice[i][j] = info[j]
            if (ice[i][j] != 0) {
                iceQueue.add(Pair(i, j))
            }
        }
    }

    val answer = melt(ice, iceQueue)
    bw.write("$answer")
    bw.newLine()
    bw.flush()
    bw.close()
}


private fun melt(ice: Array<IntArray>, queue: ArrayDeque<Pair<Int, Int>>): Int {

    var year = 0

    while(queue.isNotEmpty()) {

        val meltIce = ArrayDeque<Triple<Int, Int, Int>>()

        if (check(ice, queue)) {
            return year
        }

        while(queue.isNotEmpty()) {

            val (x, y) = queue.removeFirst()
            var count = 0
            for (i in 0 until 4) {
                val nx = x + dx[i]
                val ny = y + dy[i]

                if (nx < 0 || nx >= ice.size || ny < 0 || ny >= ice[0].size) {
                    continue
                }

                if (ice[nx][ny] == 0) {
                    count++
                }
            }
            meltIce.add(Triple(x, y, count))
        }

        while(meltIce.isNotEmpty()) {
            val (x, y, count) = meltIce.removeFirst()
            if (ice[x][y] - count > 0) {
                ice[x][y] = ice[x][y] - count
                queue.add(Pair(x, y))
            }
            else {
                ice[x][y] = 0
            }
        }

        year++
    }

    return 0
}

private fun check(ice: Array<IntArray>, queue: ArrayDeque<Pair<Int, Int>>): Boolean {

    val x = ice.size
    val y = ice[0].size

    val visited = Array(x) { BooleanArray(y) }
    var count = 0

    for (i in 0 until x) {
        for (j in 0 until y) {
            if (ice[i][j] != 0 && !visited[i][j]) {
                count++
                if (count == 2) return true
                visited[i][j] = true
                checkIce(i, j, ice, visited)
            }
        }
    }

    return false
}

private fun checkIce(sx: Int, sy: Int, ice: Array<IntArray>, visited: Array<BooleanArray>) {

    val queue = ArrayDeque<Pair<Int, Int>>()
    queue.add(Pair(sx, sy))

    while(queue.isNotEmpty()) {
        val (x, y) = queue.removeFirst()

        for (i in 0 until 4) {
            val nx = x + dx[i]
            val ny = y + dy[i]

            if (nx < 0 || nx >= ice.size || ny < 0 || ny >= ice[0].size || visited[nx][ny]) {
                continue
            }

            if (ice[nx][ny] != 0) {
                visited[nx][ny] = true
                queue.add(Pair(nx, ny))
            }
        }
    }

}